### ISSUE TITLE
Model checkpoints required for inference are different from checkpoints saved during training.

### ISSUE THREAD

**Issue by anhtuanhsgs** [visinf/acis#7]
Hello @arnike,

I tried to follow your instruction, the inference part have some errors as there is an inconsistency between checkpoint names of training and inference. . In  `cvppp_main.lua`:

  ```
local model = {}
  local actPath = paths.concat(dir, 'act.t7')
  assert(paths.filep(actPath), "Model act NOT found: " .. actPath)
  model.act = torch.load(actPath)
  model.act:evaluate()

  local decPath = paths.concat(dir, 'dec.t7')
  assert(paths.filep(decPath), "Model dec NOT found: " .. decPath)
  model.dec = torch.load(decPath)
  model.dec:evaluate()

  local ctlPath = paths.concat(dir, 'ctl.t7')
  assert(paths.filep(ctlPath), "Model ctl NOT found: " .. ctlPath)
  model.ctl = torch.load(ctlPath)
  model.ctl:evaluate()

  local ppcPath = paths.concat(dir, 'ppc.t7')
  assert(paths.filep(ppcPath), "Model ppc NOT found: " .. ppcPath)
  model.ppc = torch.load(ppcPath)
  model.ppc:evaluate()
```
_dec.t7_, _act.t7_, _ctl.t7_, _pc.t7_ are not the names of saved checkpoints during training. After finish the training, [index]_critic_opt.t7_, [index]_actor_opt.t7_, [index]_cvae_dec.t7_, [index]_cvae_enc.t7_, [index]_cvae_opt_dec.t7_, [index]_cvae_opt_enc.t7_ are name of the checkpoints of the agent.

Would you please show me how the checkpoints should be renamed for inference code?

**Comment by arnike** [comment]
Hi @anhtuanhsgs,

thanks for pointing this out!
The naming follows the following mapping:
```
actor.model -> act.t7   # (the actor, or the encoder part)
cvae_dec.t7 -> dec.t7   # (the pre-trained decoder)
control.model -> ctl.t7 # (the control network that predicts stop/continue)
preproc.model -> ppc.t7 # (the pre-processing network)
```
I usually just create symlinks to the model I want to evaluate.
It seems that you don't have `control.model` yet. It can be trained when you are satisfied with the IoU by resuming the training with additional parameters:
```
-learnControl         # train the control net
-reset_control        # (optional) ignore the control net snapshot, if found (train from scratch)
-betaCTL [MULTIPLIER] # (optional) specifies a learning rate multiplier for the parameters in the control net
```

Hope it helps.

**Comment by anhtuanhsgs** [comment]
The code below is from models/init.lua
```
--
--  Copyright (c) 2016, Facebook, Inc.
--  All rights reserved.
--
--  This source code is licensed under the BSD-style license found in the
--  LICENSE file in the root directory of this source tree. An additional grant
--  of patent rights can be found in the PATENTS file in the same directory.
--
--  Generic model creating code. For the specific ResNet model see
--  models/resnet.lua
--

require 'nn'
require 'cunn'
require 'cudnn'

local M = {}

function M.setup(netType, opt, checkpoint)
   --local model
   --if checkpoint then
   --   local modelPath = paths.concat(opt.resume, checkpoint.modelFile)
   --   assert(paths.filep(modelPath), 'Saved model not found: ' .. modelPath)
   --   print('=> Resuming model from ' .. modelPath)
   --   model = torch.load(modelPath):cuda()
   --elseif opt.retrain ~= 'none' then
   --   assert(paths.filep(opt.retrain), 'File not found: ' .. opt.retrain)
   --   print('Loading model from file: ' .. opt.retrain)
   --   model = torch.load(opt.retrain):cuda()
   --else
   --end

   print('=> Creating model from file: models/' .. netType .. '.lua')
   local model = require('models/' .. netType)(opt)
```


I tried to run as you suggested, it failed at `local model = require('models/' .. netType)(opt)`
I inspected that and it seems that `netType=control.lua` which does not exist.
From the comment control.lua can be resnet model. Could you please provide us the model.lua file?

My training option is as follow:
```
CRAYON="-crayonPort 6043"
MODEL_OPTS="-reset_control -learnControl -markov -rnn_layers 2 -rnnSize 512 -latentSize 16 -predictFg -predictAngles -pyramidImage -pyramidFg -pyramidAngles -oldControl -bottleneck fc -actorVer simple -contextType max"
TRAIN_OPTS="-learning_rate 5e-4 -lambdaKL 1e-4 -optimActor adam -learning_rate_decay 0.0 -gradClip 1 -grad ac -batch_size 8 -weightDecay 1e-5 -matchNoise 0.0"
CRITIC_OPTS="-criticVer simple -criticType fc -critic_iter 4 -gamma 0.9 -betaCritic 2 -criticNoise 0 -realNoise 0 -critic_warmup 1024 -reset_critic -nSamples 1"
LOG_OPTS="-checkpoint_after 1000 -validation_after 200 -summary_after 100 -numTestRuns 1 -numTrainRuns 1 -valTrain -nEpochs 4000"
DATA_OPTS="-dataset cvppp -seq_length 40 -seq_length_test 40 -loadLast"
```

**Comment by anhtuanhsgs** [comment]
Also, I am trying to run your models on images which each has around 60 objects on average. Images are of biomedial images with dense segmentation. The actor critic training result seems to be very poor although the decoder can capture the position and shapes. Do you have any suggestion?

**Comment by Khoa-NT** [comment]
Hi @arnike ,

Can you support me for this problem?
I follow @anhtuanhsgs and stuck here too

**Comment by Khoa-NT** [comment]
hi @arnike ,
Can you help us?

**Comment by arnike** [comment]
Hi @anhtuanhsgs, @shaolinkhoa,

> From the comment control.lua can be resnet model. Could you please provide us the model.lua file?

Sorry for the missing file, now it's in the repository. `control.lua` simply defines an MLP that is attached to the LSTM's hidden state and predicts [0, 1] indicating "stop" or "continue".

As for biomedial images: I haven't tried this. 60 objects is quite sizeable for an RNN to learn, and also rather challenging for the AC: the critic network will be asked to predict IoU/Dice a lot of iterations into the future, that's hard to learn.
What I'd try is:
1) Try if the RNN can be learned in principle. You can run the code with `opt.grad == "fixed"`: this will not use the critic net; the loss will be computed with `BCE` and the Hungarian assignment. You may need to re-adjust some of the training parameters.
2) See if AC can learn short sequences first (let's say 5). If it does, go to 4), otherwise 3).
3) The problem is with the encoder and/or LSTM and their training schedule needs to be re-adjusted to the new dataset.
4) The problem is probably the critic the network. How well can it approximate the error? 

Hope it helps,
Nikita
